File: src/render/lighter/direct.rs
use crate::geo;
use crate::object;
use crate::render;

use geo::Vec3;

use object::Intersection;
use object::Radiance;
use object::Sampler;

use render::Lighter;

pub struct Direct;

impl Direct {
    pub fn new() -> Direct {
        return Direct {};
    }
}

impl Lighter for Direct {
    fn light(
        &self,
        isect: &Intersection,
        sampler: &mut dyn Sampler,
    ) -> Radiance {
        let scene = isect.scene;
        let primitive = isect.primitive;
        let surface = &primitive.surface;
        let nrm_facing = isect.facing_normal;

        let mut rad = Radiance::ZERO;
        if let Some(light) = &primitive.light {
            rad += light.radiance_from_isect(isect);
        }
        let pnt_offset = isect.point + Vec3::from(nrm_facing) * 0.01;

        for light in &scene.lights {
            if let Some(sample) = light.sample(sampler, pnt_offset) {
                let dot = sample.direction * nrm_facing;

                if dot > 0.0 && light.test_visible(scene, &sample) {
                    let irad = sample.radiance * dot;
                    rad += irad * surface.reflected(isect, sample.direction) / sample.pdf.as_f32();
                }
            }
        }

        return rad;
    }
}

File: Render/Cpu/Impl/Lighter/Direct.cpp
#define _USE_MATH_DEFINES
#include "Render/Cpu/Impl/Lighter/Direct.hpp"

#include "Object/Scene.hpp"

#include <cmath>
#include <cfloat>

namespace Render::Cpu::Impl::Lighter {
    Math::Radiance Direct::light(const Object::Intersection &isect, Math::Sampler &sampler) const
    {
        const Object::Scene &scene = isect.scene();
        const Object::Surface &surface = isect.primitive().surface();
        const Math::Point &point = isect.point();
        const Math::Normal &nrmFacing = isect.facingNormal();

        Math::Radiance rad;
        if(isect.primitive().light()) {
            rad += isect.primitive().light()->radiance(isect);
        }

        Math::Point pntOffset = isect.point() + Math::Vector(nrmFacing) * 0.01f;

        for(const Object::Light &light : scene.lights()) {
            Object::Light::Sample sample = light.sample(sampler, pntOffset);

            float dot = sample.direction * nrmFacing;
            if(dot > 0 && light.testVisible(scene, sample)) {
                Math::Radiance irad = sample.radiance * dot;
                rad += irad * surface.reflected(isect, sample.direction) / sample.pdf;
            }
        }

        return rad;
    }
}